File: app/frontend/src/lib/api/profiles.ts
import { api, readError } from './http.ts'

export type ProfileInfo = {
  id: string
  displayName: string
  locked?: boolean
  active?: boolean
}

export type RemovedProfile = {
  id: string
  displayName: string
  removedAt: number
  expiresAt: number
}

export type ProfilesPayload = {
  activeId: string
  profiles: ProfileInfo[]
  removed?: RemovedProfile[]
}

const PROFILE_KEY = 'blombo-active-profile'
const GENERATE_KEY = 'blombo-generate'

export function storedProfileId(): string {
  try {
    return localStorage.getItem(PROFILE_KEY) || 'default'
  } catch {
    return 'default'
  }
}

export function rememberProfileId(id: string): void {
  try {
    localStorage.setItem(PROFILE_KEY, id)
  } catch {
    /* ignore */
  }
}

export function generatePersistKey(): string {
  return `${GENERATE_KEY}:${storedProfileId()}`
}

export function readGeneratePersist(): string | null {
  try {
    const key = generatePersistKey()
    let raw = localStorage.getItem(key)
    if (raw == null && storedProfileId() === 'default') {
      raw = localStorage.getItem(GENERATE_KEY)
      if (raw != null) {
        localStorage.setItem(key, raw)
      }
    }
    return raw
  } catch {
    return null
  }
}

export function writeGeneratePersist(value: string): void {
  try {
    localStorage.setItem(generatePersistKey(), value)
  } catch {
    /* ignore */
  }
}

export function removeGeneratePersist(): void {
  try {
    localStorage.removeItem(generatePersistKey())
  } catch {
    /* ignore */
  }
}

export async function getProfiles(): Promise<ProfilesPayload> {
  const res = await fetch(api('/profiles'))
  if (!res.ok) {
    throw new Error(await readError(res))
  }
  return (await res.json()) as ProfilesPayload
}

export async function createProfile(displayName: string): Promise<ProfileInfo> {
  const res = await fetch(api('/profiles'), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ displayName }),
  })
  if (!res.ok) {
    throw new Error(await readError(res))
  }
  return (await res.json()) as ProfileInfo
}

export async function renameProfile(id: string, displayName: string): Promise<ProfileInfo> {
  const res = await fetch(api(`/profiles/${encodeURIComponent(id)}`), {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ displayName }),
  })
  if (!res.ok) {
    throw new Error(await readError(res))
  }
  return (await res.json()) as ProfileInfo
}

export async function deleteProfile(id: string): Promise<void> {
  const res = await fetch(api(`/profiles/${encodeURIComponent(id)}`), { method: 'DELETE' })
  if (!res.ok) {
    throw new Error(await readError(res))
  }
}

export async function restoreProfile(id: string): Promise<ProfileInfo> {
  const res = await fetch(api(`/profiles/${encodeURIComponent(id)}/restore`), { method: 'POST' })
  if (!res.ok) {
    throw new Error(await readError(res))
  }
  return (await res.json()) as ProfileInfo
}

export async function purgeProfile(id: string): Promise<void> {
  const res = await fetch(api(`/profiles/${encodeURIComponent(id)}/purge`), { method: 'DELETE' })
  if (!res.ok) {
    throw new Error(await readError(res))
  }
}

export async function activateProfile(id: string): Promise<ProfilesPayload> {
  rememberProfileId(id)
  const res = await fetch(api(`/profiles/${encodeURIComponent(id)}/activate`), { method: 'POST' })
  if (!res.ok) {
    throw new Error(await readError(res))
  }
  return (await res.json()) as ProfilesPayload
}

File: app/backend/src/api/routes/profiles.py
from __future__ import annotations

from fastapi import APIRouter

from api.errors import ApiError
from features.profiles import service as profiles
from features.profiles.schemas import ProfileCreateIn, ProfileRenameIn

api = APIRouter()


def _raise(exc: profiles.ProfileError) -> None:
    raise ApiError(exc.code, str(exc), exc.status) from exc


@api.get("/profiles")
def get_profiles() -> dict:
    return profiles.list_profiles()


@api.post("/profiles")
def post_profile(body: ProfileCreateIn) -> dict:
    try:
        return profiles.create(body.displayName)
    except profiles.ProfileError as exc:
        _raise(exc)
        raise


@api.patch("/profiles/{ident}")
def patch_profile(ident: str, body: ProfileRenameIn) -> dict:
    try:
        return profiles.rename(ident, body.displayName)
    except profiles.ProfileError as exc:
        _raise(exc)
        raise


@api.delete("/profiles/{ident}")
def delete_profile(ident: str) -> dict:
    try:
        profiles.delete(ident)
    except profiles.ProfileError as exc:
        _raise(exc)
        raise
    return {"ok": True}


@api.post("/profiles/{ident}/restore")
def restore_profile(ident: str) -> dict:
    try:
        return profiles.restore(ident)
    except profiles.ProfileError as exc:
        _raise(exc)
        raise


@api.delete("/profiles/{ident}/purge")
def purge_profile(ident: str) -> dict:
    try:
        profiles.purge(ident)
    except profiles.ProfileError as exc:
        _raise(exc)
        raise
    return {"ok": True}


@api.post("/profiles/{ident}/activate")
def activate_profile(ident: str) -> dict:
    try:
        return profiles.activate(ident)
    except profiles.ProfileError as exc:
        _raise(exc)
        raise
